### DISCUSSION TITLE
Deoxys codebase Q&A

### DISCUSSION THREAD

**Discussion by antiyro** [madara-alliance/madara#181] [Q&A]
## Welcome to the Discussion on the Deoxys Codebase

Here, you can ask questions or directly find answers regarding the current codebase.

## Example Questions

- **What encoding do you use to store your data?**
- **Where is your synchronization with the feeder stored?**
- **Were and how do you compute commitments?**

We will make sure to respond as quickly as possible!

**Comment by antiyro** [comment]
What encoding do you use to store your data?

**Reply by jbcaron** [reply]
In our codebase, we utilize the bincode crate to serialize and deserialize our custom data types. Bincode is a binary encoding format that is highly efficient and suitable for performance-critical applications

**Reply by antiyro** [reply]
Additionaly we entirely removed [parity-scale-codec](https://github.com/paritytech/parity-scale-codec) from our codebase to avoid dealing with multiple encodings and forks. The only crate still using it is the [bonsai-trie lib](https://github.com/keep-starknet-strange/bonsai-trie), but this may change.

**Comment by antiyro** [comment]
Where is your synchronization with the feeder stored?

**Reply by antiyro** [reply]
Our syncronisation with the feeder gateway is stored under the [dp-sync](https://github.com/KasarLabs/deoxys/tree/main/crates/client/sync) crate and more precisely under [L2.rs](https://github.com/KasarLabs/deoxys/blob/main/crates/client/sync/src/l2.rs) for the Layer 2 sync. Not to be confused with the [L1.rs](https://github.com/KasarLabs/deoxys/blob/main/crates/client/sync/src/l2.rs) file which, in parallel, subscribes to the `logStateUpdate` event of Starknet verifier to listen for any verification update on the L1 (`block_hash`, `state_root`, `block_number`) and compares it with the synced state on the L2.

**Comment by antiyro** [comment]
Were and how do you compute commitments?

**Reply by antiyro** [reply]
Our commitments are computed under the [commitments](https://github.com/KasarLabs/deoxys/tree/main/crates/client/sync/src/commitments) folder in the [dp-sync](https://github.com/KasarLabs/deoxys/tree/main/crates/client/sync) crate.

- `classes.rs`: The class trie encodes the information about all existing [classes](https://docs.starknet.io/architecture-and-concepts/smart-contracts/contract-classes/) in Starknet’s state.
- `contract.rs`: The contract trie encodes the storage of all existing [contracts](https://docs.starknet.io/architecture-and-concepts/smart-contracts/contract-classes/) in Starknet's state.
- `event/transaction.rs`: Those are non-persistant tries for the txs and event commitments of each blocks, so we compute them in memory using an hashmap db

All the commitments are currently computed via the [bonsai trie lib](https://github.com/keep-starknet-strange/bonsai-trie)

**Comment by antiyro** [comment]
How are parallelization and threads handled in the code base?

**Comment by antiyro** [comment]
How and were are stored your RPC methods?

**Comment by antiyro** [comment]
I see a `mapping_db.rs` file in the dc-db crate client side, is it the same as Madara? How does your storage work?
